designs/ami-minimum-age.md

```diff
@@ -0,0 +1,70 @@
+# Ami Minimum Age
+
+This document describes a design for a new feature to allow users to specify a minimum age for AMIs.
+
+(a @jonathan-innis's idea, see [this comment](https://github.com/aws/karpenter-provider-aws/issues/5382#issuecomment-1868068193))
+
+## Background
+
+Karpenter allows users to specify TTL for nodes, which configures the maximum age of nodes in the cluster. 
+
+This is useful to ensure that nodes are replaced periodically, and helps to ensure that the cluster is running on the latest AMIs available, 
+simplifying and automating the security patching process.
+
+However, a recent issue with a released AMI (see [this issue](https://github.com/awslabs/amazon-eks-ami/issues/1551)) has highlighted that it would be useful to be able to specify a minimum age for AMIs, to ensure that nodes are not replaced with an AMI that is too new.
+
+## Proposed solution
+
+Add a new field `amiMinimumAge` to the `EC2NodeClass`
+
+```
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: my-node-class
+spec:
+  amiFamily: AL2
+  amiMinimumAge: 2w
```

Comment:
I feel like this might make more sense to be a sub-field of an AMI Selector Term. I could imagine users wanting to set a amiMinimumAge as another selection logic for each term, rather than prescribing it as an overlay over all selector terms.

Reply:
I see your point.

Just to double check, if `amiMinimumAge` was a sub-field of AMI Selector Term, it wouldn't be available to users that only want to set `amiFamily`.

Users that only need to specify  the `amiFamily` and a minimum age, would need to use a configuration like the following (assuming AL2 and eks 1.28):

```
amiSelectorTerms:
- name: "amazon-eks*node-1.28*"
   minimumAge: 5d
```
it seems to complicate a bit the simplest configuration scenario (I need the most recent AMI of the specified family, older than `minAge`), but I guess you see `amiMinimumAge` as a sort of "advanced" configuration option that warrants some more maintenance work from users? 
I'm thinking in particular at use cases when eks gets updated. Where before users that only specified `amiFamily` didn't have to upgrade their NodeClasses, now they would have to push a new NodeClasses configuration.

I'm wondering if we could do both?
The way the AMIs are discovered right now is mutually exclusive: https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/ami.go#L118

if there are `amiSelectorTerms` specified, the `amiFamily` is ignored. 

Can we adopt the same logic and have something like
```
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: my-node-class
spec:
  amiFamily: AL2
  amiMinimumAge: 2w
```

as well as
```
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: my-node-class
spec:
  amiSelectorTerms:
  - name: "amazon-eks*node-1.28*"
     minimumAge: 5d
```

and have the best of the 2 world?

Reply:
You bring up a good point, and even if there was a way to have an amiSelectorTerm, it's not _as_ simple to define it within each term. On the topic of having it in two places, I'd really opt against that, as that it may be a best of both worlds for some people, but is likely to have semantic and education confusion later.

Reply:
So we c/should keep it as first level attribute (`spec.amiMinimumAge`)?